build.zig

```diff
@@ -320,3 +328,144 @@ fn link_tracer_backend(
         },
     }
 }
+
+fn go_client(
+    b: *std.build.Builder,
+    header_generate_step: *std.build.Step,
+    mode: Mode,
+    options: *std.build.OptionsStep,
+    tracer_backend: config.TracerBackend,
+) void {
+    const build_step = b.step("go_client", "Build Go client shared library");
+
+    // Updates the generated header file:
+    const install_header = b.addInstallFile(
+        .{ .path = "src/clients/c/tb_client.h" },
+        "../src/clients/go/pkg/native/tb_client.h",
+    );
+    install_header.step.dependOn(header_generate_step);
+
+    // Zig cross-targets
+    const platforms = .{
+        "x86_64-linux",
+        "x86_64-macos",
+        "x86_64-windows",
+        "aarch64-linux",
+        "aarch64-macos",
+    };
+
+    inline for (platforms) |platform| {
+        const cross_target = CrossTarget.parse(.{ .arch_os_abi = platform, .cpu_features = "baseline" }) catch unreachable;
+
+        const lib = b.addStaticLibrary("tb_client", "src/clients/c/tb_client.zig");
+        lib.setMainPkgPath("src");
+        lib.setTarget(cross_target);
+        lib.setBuildMode(mode);
+        lib.linkLibC();
+        lib.pie = true;
+        lib.bundle_compiler_rt = true;
+
+        lib.setOutputDir("src/clients/go/pkg/native/" ++ platform);
+
+        set_cache_dir(b, platform);
+
+        lib.addOptions("tigerbeetle_build_options", options);
+        link_tracer_backend(lib, tracer_backend, cross_target);
+
+        lib.step.dependOn(&install_header.step);
+        build_step.dependOn(&lib.step);
+    }
+}
+
+fn java_client(
+    b: *std.build.Builder,
+    mode: Mode,
+    options: *std.build.OptionsStep,
+    tracer_backend: config.TracerBackend,
+) void {
+    const build_step = b.step("java_client", "Build Java client shared library");
+
+    // Zig cross-targets
+    const platforms = .{
+        "x86_64-linux-gnu",
+        "x86_64-linux-musl",
+        "x86_64-macos",
+        "aarch64-linux-gnu",
+        "aarch64-linux-musl",
+        "aarch64-macos",
+        "x86_64-windows",
+    };
```

Comment:
What is the reason for building the client with every platform instead of using Zig's default args for cross-compilation (defaulting to only current platform)?

Reply:
It's the zig replacement for the [rebuild_binaries.sh](https://github.com/tigerbeetledb/tigerbeetle/blob/f5a44d3430c7df27d9a566f8e19cab5759d45ddf/src/clients/go/scripts/rebuild_binaries.sh) script, which was deleted in this PR.
For packaging purposes (Go, Java and Dotnet), we build for all supported platforms.

Reply:
Gotcha. For development though it is a little inconvenient — if I am testing locally I probably only care about a single client. :thinking: 

What if `java_client()` (et al.):
- didn't call ` b.step("java_client", ...)` 
- only builds a single platform (passed in as an argument)
- returns the `lib`

Then in `fn build()`, create 2 steps per client: `java_client` builds the java client for the default platform (as determined by zig's build), and `java_client_all` which builds for every platform.

Reply:
It certainly can be done, but to be honest, building for all platforms didn't bother me during the development of the Java and Dotnet clients, Zig's cache works very well, and even for fresh builds it's quite fast. 

One minor issue I can point out is determining when invoking `java_client` or `java_client_all` from the target language build system:

For example:

bash
```
# Maven build calls zig java_client
(cd src/clients/java && mvn package)

# MSBuild calls zig build dotnet_client
(cd src/clients/dotnet && dotnet build)
```

It's can be easily solved by adding build arguments for Java/Dotnet builds, no problem at all, but I think it can be confusing, eg. building a java client that is not cross-platform. On the other hand, it can be done on purpose to generate smaller packages just for one specific platform 🤔.

(...)

Just my 5¢, I don't have a strong opinion about that, and I'll be happy with the result either way!

Reply:
Alright, if it isn't too slow then let's leave it as-is :+1: